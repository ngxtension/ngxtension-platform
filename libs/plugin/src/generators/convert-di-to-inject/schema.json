{
	"$schema": "http://json-schema.org/schema",
	"$id": "",
	"title": "",
	"type": "object",
	"properties": {
		"project": {
			"type": "string",
			"aliases": ["name", "projectName"],
			"description": "Project for which to convert from constructor DI to inject function."
		},
		"path": {
			"type": "string",
			"description": "Path to the component/directive/pipe you want to convert to use the inject function"
		},
		"includeReadonlyByDefault": {
			"type": "boolean",
			"aliases": ["includeReadonly", "includeReadonlyByDefault"],
			"description": "Whether to include readonly for injections by default. Default is false."
		},
		"useEsprivateFieldNotation": {
			"type": "boolean",
			"aliases": [
				"useESPrivateFieldNotation",
				"useEsprivateFieldNotation",
				"useEsPrivateFieldNotation"
			],
			"description": "Whether to replace TS private modifier with ES private field notation. Default is false."
		}
	}
}
