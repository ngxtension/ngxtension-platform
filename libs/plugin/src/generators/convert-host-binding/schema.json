{
	"$schema": "http://json-schema.org/schema",
	"$id": "ConvertHostBinding",
	"title": "",
	"type": "object",
	"properties": {
		"project": {
			"type": "string",
			"aliases": ["name", "projectName"],
			"description": "Project for which to convert from host binding to host property."
		},
		"path": {
			"type": "string",
			"description": "Path to the component/directive you want to convert to use the host property."
		}
	}
}
