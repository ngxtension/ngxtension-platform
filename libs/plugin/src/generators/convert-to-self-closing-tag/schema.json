{
	"$schema": "http://json-schema.org/schema",
	"$id": "",
	"title": "",
	"type": "object",
	"properties": {
		"project": {
			"type": "string",
			"aliases": ["name", "projectName"],
			"description": "Project for which to convert to self closing tags."
		},
		"path": {
			"type": "string",
			"description": "Path to the component/directive you want to convert to self closing tags"
		}
	}
}
