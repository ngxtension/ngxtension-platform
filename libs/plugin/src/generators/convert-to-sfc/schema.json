{
	"$schema": "http://json-schema.org/schema",
	"$id": "",
	"title": "",
	"type": "object",
	"properties": {
		"project": {
			"type": "string",
			"aliases": ["name", "projectName"],
			"description": "Project for which to convert to single file components."
		},
		"path": {
			"type": "string",
			"description": "Path to the component to convert to a single file component."
		},
		"maxInlineTemplateLines": {
			"type": "number",
			"description": "If the template has more lines than this, it will be skipped.",
			"default": 200
		},
		"maxInlineStyleLines": {
			"type": "number",
			"description": "If the styles have more lines than this, they will be skipped.",
			"default": 200
		},
		"moveStyles": {
			"type": "boolean",
			"description": "Move styles to the single file component.",
			"default": false
		}
	}
}
