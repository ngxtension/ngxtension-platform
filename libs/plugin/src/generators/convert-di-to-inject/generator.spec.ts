import { Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';

import convertDiToInjectGenerator from './generator';
import { ConvertDiToInjectGeneratorSchema } from './schema';

const filesMap = {
	notComponentNorDirective: `
	  export class MyClass {
      constructor() {}
    }
  `,
	componentNoConstructor: `
    import { Component } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {}
  `,
	componentWithDep: `
    import { Component, Optional, Attribute, ElementRef } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      constructor(
        private service: MyService,
        private readonly service2: ElementRef<HtmlImageElement>,
        @Optional() @Attribute('type') type: string,
      ) {}
    }
  `,
	componentWithDepUsage: `
    import { Component } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      constructor(
        private service: MyService
        private readonly service1: MyService1<string>
      ) {
        this.service.doSomething();
      }
    }
  `,
	componentWithDepUsageLocal: `
    import { Component } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      constructor(service: MyService) {
        service.doSomething();
        service.doSomethingElse();
      }
    }
  `,
	componentWithDepUsageLocalNoScope: `
    import { Component } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      constructor(service: MyService) {
        service.doSomething();

        someList.forEach(() => {
          // nested scope
          service.doSomething();
        });

        // use service in a function call
        someFunction(service);
      }
    }
  `,
	componentWithConstructorWithoutDep: `
    import { Component } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      constructor() {
        console.log('hello');
      }
    }`,
	componentWithDepAndInject: `
    import { Component, Inject, Attribute, Optional } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      service3: MyService3;

      constructor(
        @Attribute('type') private type: string,
        @Inject('my-service') private service: MyService,
        @Inject(MyService4) private service4: MyService4,
        @Optional() @Inject('my-service2') private service5: MyService5,
        @Inject(SOME_TOKEN) private someToken,
        private service2: MyService2,
        service3: MyService3
      ) {
        this.service3 = service3;
      }
    }
`,
	componentWithDepAndInjectAndOptions: `
    import { Component, Inject, Optional, Self } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      constructor(
        private myService: MyService,
        private elRef: ElementRef<HtmlImageElement>,
        private tplRef: TemplateRef<any>,
        private viewContainerRef: ViewContainerRef,
        service2: MyService2,
        @Inject('my-service') private service: MyService,
        @Inject(MyService4) private service4: MyService4,
        @Optional() @Inject('my-service2') private service5: MyService5,
        @Self() @Optional() private service6: MyService6,
      ) {
        myService.doSomething();

        this.service2.doSomethingElse();

        service2.doSomething();

        someList.forEach(() => {
          // nested scope
          myService.doSomething();
        });

        // use service in a function call
        someFunction(service2).test(myService);
      }
    }
`,
	componentWithDepAndInjectAndOptionsAndPrivateFields: `
    import { Component, Inject, Optional, Self } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      constructor(
        private myService: MyService,
        private elRef: ElementRef<HtmlImageElement>,
        private tplRef: TemplateRef<any>,
        private viewContainerRef: ViewContainerRef,
        service2: MyService2,
        @Inject('my-service') private service: MyService,
        @Inject(MyService4) private service4: MyService4,
        @Optional() @Inject('my-service2') private service5: MyService5,
        @Self() @Optional() private service6: MyService6,
        protected service7: MyService7,
        public service8: MyService8
      ) {
        myService.doSomething();

        this.service2.doSomethingElse();

        service2.doSomething();

        someList.forEach(() => {
          // nested scope
          myService.doSomething();
        });

        // use service in a function call
        someFunction(service2).test(myService);

        service7.getTask();
        service8.getTick();
      }

      myMethod() {
        this.myService.doCompleteTask();
        this.service7.getTask();
        this.service8.getTick();
      }

      get value() {
        return this.myService.getValue();
      }
    
      set value(theAge: number) {
        this.myService.setValue();
      }
    }
`,
	componentWithDepAndInjectAndReadonly: `
    import { Component, Inject, Attribute, Optional } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      service3: MyService3;

      constructor(
        @Attribute('type') private type: string,
        @Inject('my-service') private service: MyService,
        @Inject(MyService4) private service4: MyService4,
        @Optional() @Inject('my-service2') private service5: MyService5,
        @Inject(SOME_TOKEN) private someToken,
        private service2: MyService2,
        service3: MyService3
      ) {
        this.service3 = service3;
      }
    }
`,
	componentWithDepAssignmentsInCtor: `
    import { Component } from '@angular/core';

    @Component({
      template: ''
    })
    export class MyComponent {
      depOne$: Observable<string>;
      depTwo$: Observable<string>;
      depThree$: Observable<string>;
      depFour$: Observable<string>;

      constructor(private service: Service) {
        this.depOne$ = service.depOne$;
        this.depTwo$ = service.depTwo$;
        this.depThree$ = this.service.depThree$;
        this.depFour$ = service.depFour$;
      }
    }
`,
} as const;

// cases
// file without constructor injection or not a component/directive/pipe/injectable
// file with component but no constructor
// file with component and constructor without dependencies
// file with component and constructor with dependencies
// file with component and constructor with dependencies and @Inject() decorator on constructor
// @Inject() decorator on constructor with string literal
// file with component and constructor with dependencies that don't have type

// remove empty constructor if it's empty and has empty body
// replace the TS 'private' access modifier with ES '#private' field notation
// add the 'readonly' keyword

describe('convertDiToInjectGenerator', () => {
	let tree: Tree;
	const options: ConvertDiToInjectGeneratorSchema = {
		path: 'libs/my-file.ts',
	};

	function setup(file: keyof typeof filesMap) {
		tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
		tree.write('package.json', `{"dependencies": {"@angular/core": "17.3.0"}}`);
		tree.write(`libs/my-file.ts`, filesMap[file]);

		return () => {
			return [tree.read('libs/my-file.ts', 'utf8'), filesMap[file]];
		};
	}

	it('should not do anything if not component/directive/pipe/injectable', async () => {
		const readContent = setup('notComponentNorDirective');
		await convertDiToInjectGenerator(tree, options);
		const [updated, original] = readContent();
		expect(updated).toEqual(original);
	});

	it('should not do anything if no constructor', async () => {
		const readContent = setup('componentNoConstructor');
		await convertDiToInjectGenerator(tree, options);
		const [updated, original] = readContent();
		expect(updated).toEqual(original);
	});

	it('should convert properly', async () => {
		const readContent = setup('componentWithDep');
		await convertDiToInjectGenerator(tree, options);
		const [updated] = readContent();
		expect(updated).toMatchSnapshot();
	});

	it('should convert properly with usage', async () => {
		const readContent = setup('componentWithDepUsage');
		await convertDiToInjectGenerator(tree, options);
		const [updated] = readContent();
		expect(updated).toMatchSnapshot();
	});

	it('should convert properly with usage local', async () => {
		const readContent = setup('componentWithDepUsageLocal');
		await convertDiToInjectGenerator(tree, options);
		const [updated] = readContent();
		expect(updated).toMatchSnapshot();
	});

	it('should convert properly with usage local no scope', async () => {
		const readContent = setup('componentWithDepUsageLocalNoScope');
		await convertDiToInjectGenerator(tree, options);
		const [updated] = readContent();
		expect(updated).toMatchSnapshot();
	});

	it('should not do anything if constructor without dep', async () => {
		const readContent = setup('componentWithConstructorWithoutDep');
		await convertDiToInjectGenerator(tree, options);
		const [updated, original] = readContent();
		expect(updated).toEqual(original);
	});

	it('should not do anything if no dep', async () => {
		const readContent = setup('componentWithConstructorWithoutDep');
		await convertDiToInjectGenerator(tree, options);
		const [updated] = readContent();
		expect(updated).toMatchSnapshot();
	});

	it('should convert properly with @Inject decorator', async () => {
		const readContent = setup('componentWithDepAndInject');
		await convertDiToInjectGenerator(tree, options);
		const [updated] = readContent();
		expect(updated).toMatchSnapshot();
	});

	it('should convert properly with @Inject decorator and options', async () => {
		const readContent = setup('componentWithDepAndInjectAndOptions');
		await convertDiToInjectGenerator(tree, options);
		const [updated] = readContent();
		expect(updated).toMatchSnapshot();
	});

	it('should convert private modifier fields to native private fields', async () => {
		const readContent = setup(
			'componentWithDepAndInjectAndOptionsAndPrivateFields',
		);
		await convertDiToInjectGenerator(tree, {
			...options,
			useEsprivateFieldNotation: true,
		});
		const [updated] = readContent();
		expect(updated).toMatchSnapshot();
	});

	it('should add the readonly keyword to converted fields', async () => {
		const readContent = setup('componentWithDepAndInjectAndReadonly');
		await convertDiToInjectGenerator(tree, {
			...options,
			includeReadonlyByDefault: true,
		});
		const [updated] = readContent();
		expect(updated).toMatchSnapshot();
	});

	it('should add the readonly keyword to converted native private fields', async () => {
		const readContent = setup('componentWithDepAndInjectAndReadonly');
		await convertDiToInjectGenerator(tree, {
			...options,
			includeReadonlyByDefault: true,
			useEsprivateFieldNotation: true,
		});
		const [updated] = readContent();
		expect(updated).toMatchSnapshot();
	});

	it('should add this keyword to all assignments in constructor', async () => {
		const readContent = setup('componentWithDepAssignmentsInCtor');
		await convertDiToInjectGenerator(tree, options);
		const [updated] = readContent();
		expect(updated).toMatchSnapshot();
	});
});
