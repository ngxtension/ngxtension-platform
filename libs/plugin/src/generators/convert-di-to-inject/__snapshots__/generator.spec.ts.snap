// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertDiToInjectGenerator should add the readonly keyword to converted fields 1`] = `
"
    import { Component, inject, HostAttributeToken } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      private readonly type = inject<string>(new HostAttributeToken('type'));
      private readonly service = inject<MyService>('my-service' as any /* TODO(inject-migration): Please check if the type is correct */);
      private readonly service4 = inject(MyService4);
      private readonly service5 = inject<MyService5>('my-service2' as any /* TODO(inject-migration): Please check if the type is correct */, { optional: true });
      private readonly someToken = inject(SOME_TOKEN);
      private readonly service2 = inject(MyService2);
      service3: MyService3;

      constructor(
) {
            const service3 = inject(MyService3);

        this.service3 = service3;
      }
    }
"
`;

exports[`convertDiToInjectGenerator should add the readonly keyword to converted native private fields 1`] = `
"
    import { Component, inject, HostAttributeToken } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      readonly #type = inject<string>(new HostAttributeToken('type'));
      readonly #service = inject<MyService>('my-service' as any /* TODO(inject-migration): Please check if the type is correct */);
      readonly #service4 = inject(MyService4);
      readonly #service5 = inject<MyService5>('my-service2' as any /* TODO(inject-migration): Please check if the type is correct */, { optional: true });
      readonly #someToken = inject(SOME_TOKEN);
      readonly #service2 = inject(MyService2);
      service3: MyService3;

      constructor(
) {
            const service3 = inject(MyService3);

        this.service3 = service3;
      }
    }
"
`;

exports[`convertDiToInjectGenerator should add this keyword to all assignments in constructor 1`] = `
"
    import { Component, inject } from '@angular/core';

    @Component({
      template: ''
    })
    export class MyComponent {
      private service = inject(Service);
      depOne$: Observable<string>;
      depTwo$: Observable<string>;
      depThree$: Observable<string>;
      depFour$: Observable<string>;

      constructor() {
        this.depOne$ = this.service.depOne$;
        this.depTwo$ = this.service.depTwo$;
        this.depThree$ = this.service.depThree$;
        this.depFour$ = this.service.depFour$;
      }
    }
"
`;

exports[`convertDiToInjectGenerator should convert private modifier fields to native private fields 1`] = `
"
    import { Component, inject } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      #myService = inject(MyService);
      #elRef = inject<ElementRef<HtmlImageElement>>(ElementRef<HtmlImageElement>);
      #tplRef = inject<TemplateRef<any>>(TemplateRef<any>);
      #viewContainerRef = inject(ViewContainerRef);

      constructor(
) {
            const service2 = inject(MyService2);

        this.#myService.doSomething();

        service2.doSomethingElse();

        service2.doSomething();

        someList.forEach(() => {
          // nested scope
          this.#myService.doSomething();
        });

        // use service in a function call
        someFunction(service2).test(this.#myService);

        this.service7.getTask();
        this.service8.getTick();
      }

      #service = inject<MyService>('my-service' as any /* TODO(inject-migration): Please check if the type is correct */);
      #service4 = inject(MyService4);
      #service5 = inject<MyService5>('my-service2' as any /* TODO(inject-migration): Please check if the type is correct */, { optional: true });
      #service6 = inject(MyService6, { self: true, optional: true });
      protected service7 = inject(MyService7);
      public service8 = inject(MyService8);

      myMethod() {
        this.#myService.doCompleteTask();
        this.service7.getTask();
        this.service8.getTick();
      }

      get value() {
        return this.#myService.getValue();
      }
    
      set value(theAge: number) {
        this.#myService.setValue();
      }
    }
"
`;

exports[`convertDiToInjectGenerator should convert properly 1`] = `
"
    import { Component, ElementRef, inject, HostAttributeToken } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      private service = inject(MyService);
      private readonly service2 = inject<ElementRef<HtmlImageElement>>(ElementRef<HtmlImageElement>);

      constructor(
) {
            const type = inject<string>(new HostAttributeToken('type'), { optional: true });
}
    }
  "
`;

exports[`convertDiToInjectGenerator should convert properly with @Inject decorator 1`] = `
"
    import { Component, inject, HostAttributeToken } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      private type = inject<string>(new HostAttributeToken('type'));
      private service = inject<MyService>('my-service' as any /* TODO(inject-migration): Please check if the type is correct */);
      private service4 = inject(MyService4);
      private service5 = inject<MyService5>('my-service2' as any /* TODO(inject-migration): Please check if the type is correct */, { optional: true });
      private someToken = inject(SOME_TOKEN);
      private service2 = inject(MyService2);
      service3: MyService3;

      constructor(
) {
            const service3 = inject(MyService3);

        this.service3 = service3;
      }
    }
"
`;

exports[`convertDiToInjectGenerator should convert properly with @Inject decorator and options 1`] = `
"
    import { Component, inject } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      private myService = inject(MyService);
      private elRef = inject<ElementRef<HtmlImageElement>>(ElementRef<HtmlImageElement>);
      private tplRef = inject<TemplateRef<any>>(TemplateRef<any>);
      private viewContainerRef = inject(ViewContainerRef);

      constructor(
) {
            const service2 = inject(MyService2);

        this.myService.doSomething();

        service2.doSomethingElse();

        service2.doSomething();

        someList.forEach(() => {
          // nested scope
          this.myService.doSomething();
        });

        // use service in a function call
        someFunction(service2).test(this.myService);
      }

      private service = inject<MyService>('my-service' as any /* TODO(inject-migration): Please check if the type is correct */);
      private service4 = inject(MyService4);
      private service5 = inject<MyService5>('my-service2' as any /* TODO(inject-migration): Please check if the type is correct */, { optional: true });
      private service6 = inject(MyService6, { self: true, optional: true });
    }
"
`;

exports[`convertDiToInjectGenerator should convert properly with usage 1`] = `
"
    import { Component, inject } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      private service = inject(MyService);
      private readonly service1 = inject<MyService1<string>>(MyService1<string>);

      constructor(
) {
        this.service.doSomething();
      }
    }
  "
`;

exports[`convertDiToInjectGenerator should convert properly with usage local 1`] = `
"
    import { Component, inject } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      constructor() {
            const service = inject(MyService);

        service.doSomething();
        service.doSomethingElse();
      }
    }
  "
`;

exports[`convertDiToInjectGenerator should convert properly with usage local no scope 1`] = `
"
    import { Component, inject } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      constructor() {
            const service = inject(MyService);

        service.doSomething();

        someList.forEach(() => {
          // nested scope
          service.doSomething();
        });

        // use service in a function call
        someFunction(service);
      }
    }
  "
`;

exports[`convertDiToInjectGenerator should not do anything if no dep 1`] = `
"
    import { Component } from '@angular/core';
    @Component({
      template: ''
    })
    export class MyComponent {
      constructor() {
        console.log('hello');
      }
    }"
`;
